SCollection of recursive types produces NotSerializableException

I'm new to this project, so please bear with me.

I have a `SCollection` with a type referring to a recursive type, it produces this exception when I `flatMap` over it

```
Caused by: java.io.NotSerializableException: PipelineOptions objects are not serializable and should not be embedded into transforms (did you capture a PipelineOptions object in a field or in an anonymous class?). Instead, if you're using a DoFn, access PipelineOptions at runtime via ProcessContext/StartBundleContext/FinishBundleContext.getPipelineOptions(), or pre-extract necessary fields from PipelineOptions at pipeline construction time.
        at org.apache.beam.sdk.options.ProxyInvocationHandler.writeObject(ProxyInvocationHandler.java:176)
```

I reproduced the error in the test for derivation of recursive encoders, please see the linked PR.

This is the problematic path, if it's not obvious to you:

* `com.spotify.scio.values.PCollectionWrapper.parDo` materializes the coder, [link](https://github.com/spotify/scio/blob/master/scio-core/src/main/scala/com/spotify/scio/values/PCollectionWrapper.scala#L58). This gives a `com.spotify.scio.coders.WrappedBCoder`.
* [`CoderTranslation.toProto`](https://github.com/apache/beam/blob/master/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CoderTranslation.java#L82) does not recognize `WrappedBCoder` and to triggers serialization through [`toCustomCoder`](https://github.com/apache/beam/blob/master/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CoderTranslation.java#L113)

Please note that this path is not caused by the recursive type, but the recursive types fails to serialize with the error above.

--- a/scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala
@@ -167,26 +167,31 @@ private final case class DisjunctionCoder[T, Id](
     }
 }
 
-private[scio] final class RefCoder[T](val typeName: String, c: => BCoder[T]) extends BCoder[T] {
-  def value: BCoder[T] = c
+private[scio] final case class RefCoder[T](val typeName: String, var coder: BCoder[T])
+    extends BCoder[T] {
+  def setImpl(_c: BCoder[T]): Unit = coder = _c
 
-  def decode(inStream: InputStream): T = c.decode(inStream)
-  def encode(value: T, outStream: OutputStream): Unit = c.encode(value, outStream)
-  def getCoderArguments(): java.util.List[_ <: BCoder[_]] = c.getCoderArguments()
-  def verifyDeterministic(): Unit = c.verifyDeterministic()
+  private def check[A](t: => A): A = {
+    require(coder != null, s"Coder implementation should not be null in ${this}")
+    t
+  }
 
-  override def consistentWithEquals(): Boolean = c.consistentWithEquals()
-  override def structuralValue(value: T): AnyRef =
+  def decode(inStream: InputStream): T = check { coder.decode(inStream) }
+  def encode(value: T, outStream: OutputStream): Unit = check { coder.encode(value, outStream) }
+  def getCoderArguments(): java.util.List[_ <: BCoder[_]] = check { coder.getCoderArguments() }
+  def verifyDeterministic(): Unit = check { coder.verifyDeterministic() }
+
+  override def consistentWithEquals(): Boolean = check { coder.consistentWithEquals() }
+  override def structuralValue(value: T): AnyRef = check {
     if (consistentWithEquals()) {
       value.asInstanceOf[AnyRef]
     } else {
-      c.structuralValue(value)
+      coder.structuralValue(value)
     }
-}
+  }
 
-private[scio] object RefCoder {
-  def apply[T](t: String, c: => BCoder[T]): RefCoder[T] = new RefCoder[T](t, c)
-  def unapply[T](c: RefCoder[T]): Option[(String, BCoder[T])] = Option((c.typeName, c.value))
+  override def toString(): String =
+    s"RefCoder($typeName, ${if (coder == null) "null" else "<coder ref>"})"
 }
 
 // XXX: Workaround a NPE deep down the stack in Beam

--- a/scio-core/src/main/scala/com/spotify/scio/coders/CoderMaterializer.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/coders/CoderMaterializer.scala
@@ -19,6 +19,7 @@ package com.spotify.scio.coders
 
 import org.apache.beam.sdk.coders.{CoderRegistry, KvCoder, NullableCoder, Coder => BCoder}
 import org.apache.beam.sdk.options.{PipelineOptions, PipelineOptionsFactory}
+import scala.collection.concurrent.TrieMap
 
 object CoderMaterializer {
   import com.spotify.scio.ScioContext
@@ -42,13 +43,13 @@ object CoderMaterializer {
     r: CoderRegistry,
     o: PipelineOptions,
     coder: Coder[T]
-  ): BCoder[T] = beamImpl(r, o, coder, Map.empty)
+  ): BCoder[T] = beamImpl(r, o, coder, TrieMap.empty)
 
   final private def beamImpl[T](
     r: CoderRegistry,
     o: PipelineOptions,
     coder: Coder[T],
-    refs: Map[String, RefCoder[_]]
+    refs: TrieMap[String, RefCoder[_]]
   ): BCoder[T] = {
     coder match {
       // #1734: do not wrap native beam coders
@@ -87,7 +88,11 @@ object CoderMaterializer {
         refs
           .get(t)
           .getOrElse {
-            lazy val bc: BCoder[T] = beamImpl(r, o, c, refs + (t -> RefCoder(t, bc)))
+            // #2269: _REALLY_ support recursive ADTs
+            val bc = RefCoder[T](t, null) // placeholder
+            refs.put(t, bc)
+            val actualImpl = beamImpl(r, o, c, refs)
+            bc.setImpl(actualImpl)
             bc
           }
           .asInstanceOf[BCoder[T]]

--- a/scio-test/src/test/scala/com/spotify/scio/coders/CoderTest.scala
+++ b/scio-test/src/test/scala/com/spotify/scio/coders/CoderTest.scala
@@ -457,14 +457,15 @@ final class CoderTest extends FlatSpec with Matchers {
   }
 
   it should "support derivation of recursive types" in {
-    case class SampleField(name: String, fieldType: SampleFieldType)
-    sealed trait SampleFieldType
-    case object IntegerType extends SampleFieldType
-    case object StringType extends SampleFieldType
-    case class RecordType(fields: List[SampleField]) extends SampleFieldType
+    import RecursiveCase._
 
-    noException should be thrownBy SerializableUtils.serializeToByteArray(CoderMaterializer.beamWithDefault(implicitly[Coder[Top]]))
-    noException should be thrownBy SerializableUtils.serializeToByteArray(CoderMaterializer.beamWithDefault(implicitly[Coder[SampleFieldType]]))
+    noException should be thrownBy
+      SerializableUtils.serializeToByteArray(CoderMaterializer.beamWithDefault(Coder[Top]))
+
+    noException should be thrownBy
+      SerializableUtils.serializeToByteArray(
+        CoderMaterializer.beamWithDefault(Coder[SampleFieldType])
+      )
 
     "Coder[SampleField]" should compile
     // deriving this coder under 2.11 will fail
@@ -481,4 +482,14 @@ final class CoderTest extends FlatSpec with Matchers {
       )
     ) coderShould roundtrip()
   }
+}
+
+object RecursiveCase {
+  case class SampleField(name: String, fieldType: SampleFieldType)
+  sealed trait SampleFieldType
+  case object IntegerType extends SampleFieldType
+  case object StringType extends SampleFieldType
+  case class RecordType(fields: List[SampleField]) extends SampleFieldType
+
+  implicit val coderSampleFieldType = Coder.gen[SampleField]
 }

--- a/scio-test/src/test/scala/com/spotify/scio/coders/CoderTest.scala
+++ b/scio-test/src/test/scala/com/spotify/scio/coders/CoderTest.scala
@@ -25,6 +25,7 @@ import org.apache.beam.sdk.coders.Coder.NonDeterministicException
 import org.apache.beam.sdk.options.{PipelineOptions, PipelineOptionsFactory}
 import org.scalactic.Equality
 import org.scalatest.{FlatSpec, Matchers}
+import org.apache.beam.sdk.util.SerializableUtils
 
 import scala.collection.JavaConverters._
 import scala.collection.{mutable => mut}
@@ -461,6 +462,9 @@ final class CoderTest extends FlatSpec with Matchers {
     case object IntegerType extends SampleFieldType
     case object StringType extends SampleFieldType
     case class RecordType(fields: List[SampleField]) extends SampleFieldType
+
+    noException should be thrownBy SerializableUtils.serializeToByteArray(CoderMaterializer.beamWithDefault(implicitly[Coder[Top]]))
+    noException should be thrownBy SerializableUtils.serializeToByteArray(CoderMaterializer.beamWithDefault(implicitly[Coder[SampleFieldType]]))
 
     "Coder[SampleField]" should compile
     // deriving this coder under 2.11 will fail

